src/core/transaction.zig

```diff
@@ -231,16 +267,220 @@ pub const CompiledInstruction = struct {
     }
 };
 
-test "core.transaction: tmp" {
+pub const AccountMeta = struct {
+    pubkey: Pubkey,
+    is_signer: bool,
+    is_writable: bool,
+
+    pub fn newMutable(pubkey: Pubkey, is_signer: bool) AccountMeta {
+        return .{
+            .pubkey = pubkey,
+            .is_signer = is_signer,
+            .is_writable = true,
+        };
+    }
+
+    pub fn newImmutable(pubkey: Pubkey, is_signer: bool) AccountMeta {
+        return .{
+            .pubkey = pubkey,
+            .is_signer = is_signer,
+            .is_writable = false,
+        };
+    }
+};
+
+pub const CompiledKeys = struct {
+    maybe_payer: ?Pubkey,
+    key_meta_map: std.AutoArrayHashMap(Pubkey, CompiledKeyMeta),
+
+    pub fn compile(allocator: std.mem.Allocator, instructions: []Instruction, maybe_payer: ?Pubkey) !CompiledKeys {
+        var key_meta_map = std.AutoArrayHashMap(Pubkey, CompiledKeyMeta).init(allocator);
+        for (instructions) |instruction| {
+            const instruction_meta_gopr = try key_meta_map.getOrPut(instruction.program_id);
+            if (!instruction_meta_gopr.found_existing) {
+                instruction_meta_gopr.value_ptr.* = CompiledKeyMeta.default();
+            }
+            instruction_meta_gopr.value_ptr.*.is_invoked = true;
+
+            for (instruction.accounts) |account_meta| {
+                const account_meta_gopr = try key_meta_map.getOrPut(account_meta.pubkey);
+                if (!account_meta_gopr.found_existing) {
+                    account_meta_gopr.value_ptr.* = CompiledKeyMeta.default();
+                }
+                account_meta_gopr.value_ptr.*.is_signer = account_meta_gopr.value_ptr.*.is_signer or account_meta.is_signer;
+                account_meta_gopr.value_ptr.*.is_writable = account_meta_gopr.value_ptr.*.is_writable or account_meta.is_writable;
+            }
+
+            if (maybe_payer) |payer| {
+                const payer_meta_gopr = try key_meta_map.getOrPut(payer);
+                if (!payer_meta_gopr.found_existing) {
+                    payer_meta_gopr.value_ptr.* = CompiledKeyMeta.default();
+                }
+                payer_meta_gopr.value_ptr.*.is_signer = true;
+                payer_meta_gopr.value_ptr.*.is_writable = true;
+            }
+        }
+        return .{ .maybe_payer = maybe_payer, .key_meta_map = key_meta_map };
+    }
+
+    /// Creates message header and account keys from the compiled keys.
+    /// Account keys memory is allocated and owned by the caller.
+    /// TODO: Depending on whether the order of account keys is important, the code could be
+    /// optimized by simply counting the key types and appending them to account_keys direclty.
```

Comment:
On this note: even if the order does matter, based on the fact that the message header uses `u8`s to represent the number of involved accounts, if we can assume we have an upper bound of maybe `255 + 255 + 255 = 765` accounts, it would likely be in scope to init an arraylist with `capacity = self.key_meta_map.count()`, and `insert` the keys at a tracked offset, ie:
```zig
const account_keys_buf = try allocator.alloc(Pubkey, self.key_meta_map.count() + @intFromBool(self.maybe_payer != null));
errdefer allocator.free(account_keys_buf);

var account_keys = std.ArrayListUnmanaged(Pubkey).initBuffer(account_keys_buf);

if (self.maybe_payer) |payer| {
    _ = self.key_meta_map.swapRemove(payer);
    account_keys.appendAssumeCapacity(payer);
}

var writable_signers_end: usize = 1;
var readonly_signers_end: usize = 0;
var writable_non_signers_end: usize = 0;
// readonly_non_signers_end = account_keys.items.len always

for (self.key_meta_map.keys(), self.key_meta_map.values()) |key, meta| {
    if (meta.is_signer and meta.is_writable) {
        account_keys.insertAssumeCapacity(writable_signers_end, key);
        writable_signers_end += 1;
        readonly_signers_end += 1;
        writable_non_signers_end += 1;
    } else if (meta.is_signer and !meta.is_writable) {
        account_keys.insertAssumeCapacity(readonly_signers_end, key);
        readonly_signers_end += 1;
        writable_non_signers_end += 1;
    } else if (!meta.is_signer and meta.is_writable) {
        account_keys.insertAssumeCapacity(writable_non_signers_end, key);
        writable_non_signers_end += 1;
    } else if (!meta.is_signer and !meta.is_writable) {
        account_keys.appendAssumeCapacity(readonly_non_signers_end, key);
    } else unreachable;
}

std.debug.assert(account_keys.len == account_keys_buf.len);
```
^ code equivalent of a scrawl on a napkin, not tested

lmk your thoughts.

Reply:
Yeah makes sense, good idea. One question, is there a material difference between using initialising a buffer and then creating an unmanaged array list over just creating a managed array list with init capacity?

Reply:
Ahh I guess then we can just return the allocated buff without calling toOwnedSlice

Reply:
e165699